File: apps/web/src/stores/chat-room-store.ts
import { create } from 'zustand'
import { persist } from 'zustand/middleware'
import { ChatRoom, chatRoomApi } from '@/lib/agent-api'

const STORAGE_KEY = 'selectedChatRoomId'

function readUserId(rawValue: string | null): string | null {
  if (!rawValue) return null

  try {
    const parsed = JSON.parse(rawValue) as { id?: unknown; state?: { user?: { id?: unknown } | null } }
    if (typeof parsed.id === 'string') return parsed.id
    if (typeof parsed.state?.user?.id === 'string') return parsed.state.user.id
    return null
  } catch {
    return null
  }
}

function getCurrentUserIdFromStorage(): string | null {
  if (typeof localStorage === 'undefined') return null

  return readUserId(localStorage.getItem('auth_user'))
    ?? readUserId(localStorage.getItem('auth-storage'))
}

interface LastMessage {
  id: string
  content: string
  time: string
  isHuman: boolean
  userId: string | null
  agentId: string | null
  user: {
    id: string
    username: string
  } | null
  agent: {
    id: string
    name: string
  } | null
}

interface ChatRoomStore {
  chatRooms: ChatRoom[]
  selectedRoomId: string | null
  chatRoomsCacheUserId: string | null
  setChatRooms: (rooms: ChatRoom[]) => void
  selectRoom: (id: string) => void
  addRoom: (room: ChatRoom) => void
  removeRoom: (id: string) => void
  loadChatRooms: () => Promise<void>
  updateRoomLastMessage: (chatRoomId: string, lastMessage: LastMessage | null) => void
}

export const useChatRoomStore = create<ChatRoomStore>()(
  persist(
    (set, get) => ({
      chatRooms: [],
      selectedRoomId: localStorage.getItem(STORAGE_KEY),
      chatRoomsCacheUserId: getCurrentUserIdFromStorage(),

      setChatRooms: (rooms: ChatRoom[]) => {
        set({
          chatRooms: rooms,
          chatRoomsCacheUserId: getCurrentUserIdFromStorage(),
        })
      },

      selectRoom: (id: string) => {
        if (id) {
          localStorage.setItem(STORAGE_KEY, id)
          set({ selectedRoomId: id })
          return
        }

        localStorage.removeItem(STORAGE_KEY)
        set({ selectedRoomId: null })
      },

      addRoom: (room: ChatRoom) => {
        set((state) => ({
          chatRooms: state.chatRooms.some((item) => item.id === room.id)
            ? state.chatRooms.map((item) => item.id === room.id ? room : item)
            : [...state.chatRooms, room],
        }))
      },

      removeRoom: (id: string) => {
        set((state) => ({
          chatRooms: state.chatRooms.filter((room) => room.id !== id),
          selectedRoomId: state.selectedRoomId === id ? null : state.selectedRoomId,
        }))
        if (get().selectedRoomId === null) {
          localStorage.removeItem(STORAGE_KEY)
        }
      },

      loadChatRooms: async () => {
        const currentUserId = getCurrentUserIdFromStorage()
        if (get().chatRoomsCacheUserId !== currentUserId) {
          localStorage.removeItem(STORAGE_KEY)
          set({
            chatRooms: [],
            selectedRoomId: null,
            chatRoomsCacheUserId: currentUserId,
          })
        }

        const response = await chatRoomApi.getAll()
        if (response.success && response.data) {
          set({
            chatRooms: response.data,
            chatRoomsCacheUserId: currentUserId,
          })
        }
      },

      updateRoomLastMessage: (chatRoomId: string, lastMessage: LastMessage | null) => {
        set((state) => {
          const roomIndex = state.chatRooms.findIndex(room => room.id === chatRoomId)
          if (roomIndex === -1) return state

          const updatedRooms = [...state.chatRooms]
          // 更新 lastMessage 和活动时间，渲染时会按时间排序
          updatedRooms[roomIndex] = {
            ...updatedRooms[roomIndex],
            lastMessage,
            updatedAt: lastMessage?.time ?? new Date().toISOString(),
          }

          return { chatRooms: updatedRooms }
        })
      },
    }),
    {
      name: 'chat-room-storage',
      version: 1,
      migrate: (persistedState) => {
        const persisted = persistedState as Partial<ChatRoomStore> | undefined
        return {
          selectedRoomId: persisted?.selectedRoomId,
          chatRoomsCacheUserId: persisted?.chatRoomsCacheUserId,
        }
      },
      merge: (persistedState, currentState) => {
        const persisted = persistedState as Partial<ChatRoomStore> | undefined
        const currentUserId = getCurrentUserIdFromStorage()

        return {
          ...currentState,
          ...persisted,
          chatRooms: [],
          chatRoomsCacheUserId: currentUserId,
        }
      },
      partialize: (state) => ({
        selectedRoomId: state.selectedRoomId,
        chatRoomsCacheUserId: state.chatRoomsCacheUserId,
      }),
    }
  )
)

File: apps/mobile/lib/stores/chat_room_store.dart
import 'package:flutter/foundation.dart';
import '../models/models.dart';
import '../services/api_client.dart';

/// 聊天室状态管理
class ChatRoomStore extends ChangeNotifier {
  List<ChatRoom> _chatRooms = [];
  String? _selectedRoomId;
  bool _loading = false;

  List<ChatRoom> get chatRooms => _chatRooms;
  String? get selectedRoomId => _selectedRoomId;
  bool get loading => _loading;

  /// 加载聊天室列表
  Future<void> loadChatRooms() async {
    _loading = true;
    notifyListeners();

    try {
      final response = await ChatRoomApi.getAll();
      if (response.success && response.data != null) {
        _chatRooms = response.data!;
      }
      _loading = false;
      notifyListeners();
    } catch (e) {
      _loading = false;
      notifyListeners();
    }
  }

  /// 选择聊天室
  void selectRoom(String? id) {
    _selectedRoomId = id;
    notifyListeners();
  }

  /// 添加聊天室
  void addRoom(ChatRoom room) {
    _chatRooms = [room, ..._chatRooms];
    notifyListeners();
  }

  /// 移除聊天室
  void removeRoom(String id) {
    _chatRooms = _chatRooms.where((room) => room.id != id).toList();
    if (_selectedRoomId == id) {
      _selectedRoomId = null;
    }
    notifyListeners();
  }

  /// 更新聊天室
  void updateRoom(ChatRoom room) {
    _chatRooms = _chatRooms.map((r) => r.id == room.id ? room : r).toList();
    notifyListeners();
  }

  /// 创建聊天室
  Future<Map<String, dynamic>> createRoom(String name, String? description) async {
    try {
      final response = await ChatRoomApi.create(
        name: name,
        description: description,
      );

      if (response.success && response.data != null) {
        _chatRooms = [response.data!, ..._chatRooms];
        notifyListeners();
        return {'success': true};
      }

      return {'success': false, 'error': response.error ?? '创建失败'};
    } catch (e) {
      return {'success': false, 'error': '网络错误'};
    }
  }

  /// 删除聊天室
  Future<void> deleteRoom(String id) async {
    try {
      await ChatRoomApi.delete(id);
      _chatRooms = _chatRooms.where((room) => room.id != id).toList();
      if (_selectedRoomId == id) {
        _selectedRoomId = null;
      }
      notifyListeners();
    } catch (e) {
      // 忽略删除错误
    }
  }

  /// 设置聊天室列表
  void setChatRooms(List<ChatRoom> rooms) {
    _chatRooms = rooms;
    notifyListeners();
  }

  /// 获取指定聊天室
  ChatRoom? getRoomById(String id) {
    return _chatRooms.firstWhereOrNull((r) => r.id == id);
  }
}

/// List 扩展
extension ListExtension<T> on List<T> {
  T? firstWhereOrNull(bool Function(T) test) {
    for (final element in this) {
      if (test(element)) return element;
    }
    return null;
  }
}